### ISSUE TITLE
ExcludeFromUninstall issue in Remove-MSIApplications function

### ISSUE THREAD

**Issue by yulasinio** [PSAppDeployToolkit/PSAppDeployToolkit#216]
I'm trying to exclude multiple versions from being uninstalled in Remove-MSIApplications but it's not working for me.
I don't know if my array is wrong or this is how PS process the elements individually.

> Remove-MSIApplications -Name 'Test Client' -ExcludeFromUninstall @(@('DisplayVersion', '4.0.0.31731', 'Exact'),@('DisplayVersion', '4.0.1.33218', 'Exact'),@('DisplayVersion', '5.0.0.7814', 'Exact'),@('DisplayVersion', '6.0.0.34325', 'Exact')) -Exact>

On a quick debug I found that after ForEach line 2531 the $Exclude variable becomes a String and the comparison returns false.

> [DBG]: PS C:\WINDOWS\system32>> $Exclude | gm
> TypeName: System.String
> [DBG]: PS C:\WINDOWS\system32>> $Exclude[0]
> DisplayVersion
> [DBG]: PS C:\WINDOWS\system32>> $Exclude[0][2]
> s
> [DBG]: PS C:\WINDOWS\system32>> $Exclude[0][2] -eq 'Exact'
> False

The exclusion works if only one element is present in the array. $Exclude is System.Object this time.

> Remove-MSIApplications -Name 'Test Client' -ExcludeFromUninstall @(,,@('DisplayVersion', '6.0.2.13700', 'Exact'))>
> 
> [DBG]: PS C:\WINDOWS\system32>> $Exclude | gm
> TypeName: System.Object[]
> [DBG]: PS C:\WINDOWS\system32>> $Exclude[0]
> DisplayVersion
> 4.0.0.31731
> Exact
> [DBG]: PS C:\WINDOWS\system32>> $Exclude[0][2]
> Exact
> [DBG]: PS C:\WINDOWS\system32>> $Exclude[0][2] -eq 'Exact'
> True

As a quick workaround I've added an If statement to check how many elements I have in the array and if more than one is present select the correct element for comparison.
I do think that this is just a workaround and a better of of doing this is to construct objects to work with instead of the arrays.
Similar behavior is mentioned in the comment section of this blog post http://blogs.msdn.com/b/powershell/archive/2007/01/23/array-literals-in-powershell.aspx

> # Filter the results from Get-InstalledApplication to remove those that should never be uninstalled
> 
> If (($null -ne $ExcludeFromUninstall) -and ($ExcludeFromUninstall.Count)) {
>     If ($ExcludeFromUninstall.Count -eq '1') {
>         ForEach ($Exclude in $ExcludeFromUninstall) {
>             If ($Exclude[0][2] -eq 'RegEx') {
>                 If ($installedApplication.($Exclude[0][0]) -match [regex]::Escape($Exclude[0][1])) {
>                     [boolean]$addAppToRemoveList = $false
>                     Write-Log -Message "Skipping removal of application [$($installedApplication.DisplayName) $($installedApplication.Version)] because of regex match against [-ExcludeFromUninstall] criteria." -Source ${CmdletName}
>                 }
>             }
>             ElseIf ($Exclude[0][2] -eq 'WildCard') {
>                 If ($installedApplication.($Exclude[0][0]) -like $Exclude[0][1]) {
>                     [boolean]$addAppToRemoveList = $false
>                     Write-Log -Message "Skipping removal of application [$($installedApplication.DisplayName) $($installedApplication.Version)] because of wildcard match against [-ExcludeFromUninstall] criteria." -Source ${CmdletName}
>                 }
>             }
>             ElseIf ($Exclude[0][2] -eq 'Exact') {
>                 If ($installedApplication.($Exclude[0][0]) -eq $Exclude[0][1]) {
>                     [boolean]$addAppToRemoveList = $false
>                     Write-Log -Message "Skipping removal of application [$($installedApplication.DisplayName) $($installedApplication.Version)] because of exact match against [-ExcludeFromUninstall] criteria." -Source ${CmdletName}
>                 }
>             }
>         }
>     }
>     Else {
>         ForEach ($Exclude in $ExcludeFromUninstall) {
>             If ($Exclude[2] -eq 'RegEx') {
>                 If ($installedApplication.($Exclude[0]) -match [regex]::Escape($Exclude[1])) {
>                     [boolean]$addAppToRemoveList = $false
>                     Write-Log -Message "Skipping removal of application [$($installedApplication.DisplayName) $($installedApplication.Version)] because of regex match against [-ExcludeFromUninstall] criteria." -Source ${CmdletName}
>                 }
>             }
>             ElseIf ($Exclude[2] -eq 'WildCard') {
>                 If ($installedApplication.($Exclude[0]) -like $Exclude[1]) {
>                     [boolean]$addAppToRemoveList = $false
>                     Write-Log -Message "Skipping removal of application [$($installedApplication.DisplayName) $($installedApplication.Version)] because of wildcard match against [-ExcludeFromUninstall] criteria." -Source ${CmdletName}
>                 }
>             }
>             ElseIf ($Exclude[2] -eq 'Exact') {
>                 If ($installedApplication.($Exclude[0]) -eq $Exclude[1]) {
>                     [boolean]$addAppToRemoveList = $false
>                     Write-Log -Message "Skipping removal of application [$($installedApplication.DisplayName) $($installedApplication.Version)] because of exact match against [-ExcludeFromUninstall] criteria." -Source ${CmdletName}
>                 }
>             }
>         }
>     }
> }

**Comment by PLogFinder** [comment]
For now, I think the code should be:
```
ForEach ($Exclude in $ExcludeFromUninstall) {
    If ($Exclude[2] -eq 'RegEx') {
        If ($installedApplication.($Exclude[0]) -match [regex]::Escape($Exclude[1])) { ... }
```

... because  the ForEach already de-constructs the outer array dimension from $ExcludeFromUninstall, and in $Exclude we are left with a simple array of 3 elements:  [0]: property,   [1]: value,   [2]: match-type (exact, wildcard, regex)
So, I think it's a bug in the  'main' code module.  I'd just take out the first [0] in the $Exclude[0][_n_] variable references.
I'd also take the first [0] out of the $Filter[0][_n_] references for $FilterApplication iteration blocks above the exclude blocks.
If I was about to rely on that function, using the filter or exclude parameters, I'd certainly edit the script **AppDeployToolkitMain.ps1** in the package I was deploying.  And I hope it gets fixed in the next release.

**Comment by yulasinio** [comment]
> So,  I think it's a bug in the 'main' code module. I'd just take out the first [0] in the $Exclude[0][n] variable references.

If you remove the [0] it will work for multiple elements passed to -ExcludeFromUninstall but will not work for single exclusions.
That's why my workaround it's using [0][n] for single exclusions and [n] for multiple exclusions.

**Comment by PLogFinder** [comment]
Did you test a single exclusion with the extra commas as in the documentation (to force nesting)?
-ExcludeFromUninstall @(,,@('DisplayName', 'Java 8 Update 45', 'RegEx'))

or just:
-ExcludeFromUninstall ,,('DisplayName', 'Java 8 Update 45', 'RegEx')

(if it's decided that it's better to not ask a toolkit user to deal with the nesting then I can go with your solution)

**Comment by PLogFinder** [comment]
What if you use:

`Remove-MSIApplications -Name 'Test Client' -ExcludeFromUninstall ,,('DisplayVersion', '4.0.0.31731', 'Exact'),('DisplayVersion', '4.0.1.33218', 'Exact'),('DisplayVersion', '5.0.0.7814', 'Exact'),('DisplayVersion', '6.0.0.34325', 'Exact') -Exact`

**Comment by yulasinio** [comment]
Yes, for single exclusions we are using the extra commas to create the array:
`Remove-MSIApplications -Name 'Test App' -FilterApplication @(,,@('DisplayVersion', '1.1.0.11', 'Exact'))`
I'll try your suggestions in a minute and return with the results.

**Comment by PLogFinder** [comment]
If  2 sub-arrays, set them:   
**`('ab', 'cd', 'ef'), ('12','34','56')`**
(that is, 2-dimensional array, 2 x 3   elements)

If 1 sub-array, set it:  
**`,('ab', 'cd', 'ef')`**
(that is, still a 2-dimensional array, 1 x 3   elements)

Then only need [0], [1] and [2] (without the double [0][0] etc) to get the values within the foreach loops.

I think the @( ) thing is confusing in the function call and the in-line help doco.
Comma is 'array' operator and will force the 2 dimensional nesting.
@( ) is [array] cast, but will not force nesting.

So call would look like this for multiple exclusions:
**`Remove-MSIApplications -Name 'Test Client' -ExcludeFromUninstall ('DisplayVersion', '4.0.0.31731', 'Exact'),('DisplayVersion', '4.0.1.33218', 'Exact'),('DisplayVersion', '5.0.0.7814', 'Exact'),('DisplayVersion', '6.0.0.34325', 'Exact') -Exact`**

or (if only one exclusion):

**`Remove-MSIApplications -Name 'Test Client' -ExcludeFromUninstall ,('DisplayVersion', '6.0.0.34325', 'Exact') -Exact`**

**Comment by yulasinio** [comment]
OK, that worked.
I've modified the main script and removed the [0] from $Filter and $Exclude and only left the [n].

For the single exclusion you need to add another set of parenthesis or it won't execute.
`Remove-MSIApplications -Name 'Test Client' -ExcludeFromUninstall (,('DisplayVersion', '6.0.0.34325', 'Exact')) -Exact`

The syntax for multiple exclusions is correct.

**Comment by PLogFinder** [comment]
That's good news.  Good working with you!

**Comment by amotaz** [comment]
This issue is resolved in 3.6.9 using PLogFinder's code. please let me know if you still have issues.
